### PR TITLE
2318 capture all the errors on transactions

### LINKED ISSUE #2318: Capture all the errors on transactions

### PR DESCRIPTION
**Pull request by Jayashsatolia403**
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/dbs/event/transaction.go [hunk 1]
```diff
@@ -134,3 +144,54 @@ func (edb *EventDb) GetTransactionsForBlocks(blockStart, blockEnd int64) ([]Tran
...
+	var transactions []Transaction
+
+	// created_at for last day from now
+	lastDay := time.Now().AddDate(0, 0, -1)
+	// convert to string
+	lastDayString := lastDay.Format("2006-01-02 15:04:05")
+
+	logging.Logger.Info("lastDayString", zap.Any("lastDayString", lastDayString), zap.Any("lastDay", lastDay), zap.Any("time.Now()", time.Now()))
+
+	// read all the transactions from the transaction table where status is 2 till last day
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/transaction.go]: We are running this update every 60 seconds, so we don't need to get all from last day on every query, which could cause performance issue. So perhaps from last 2 minutes is enough.

File: code/go/0chain.net/smartcontract/dbs/event/transaction.go [hunk 2]
```diff
@@ -134,3 +144,54 @@ func (edb *EventDb) GetTransactionsForBlocks(blockStart, blockEnd int64) ([]Tran
...
+
+	err := edb.Get().Model(&Transaction{}).Select("output_hash").Where("status = ? and created_at > ?", 2, lastDayString).Group("output_hash").Find(&transactions)
+
+	if err.Error != nil {
+		logging.Logger.Error("Error while reading transactions from transaction table", zap.Any("error", err.Error))
+		return err.Error
+	}
+
+	logging.Logger.Info("jayashtransactionsLen", zap.Any("transactions", len(transactions)))
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/transaction.go]: Clean up the logs before merging the PR, could be `logging.Logger.Info("update transaction errors, txn len"...)`.

File: code/go/0chain.net/smartcontract/dbs/event/transaction.go [hunk 3]
```diff
@@ -134,3 +144,54 @@ func (edb *EventDb) GetTransactionsForBlocks(blockStart, blockEnd int64) ([]Tran
...
+
+	for _, transaction := range transactions {
+		// insert the transaction in the transaction error table
+
+		logging.Logger.Info("jayashTransaction : ", zap.Any("transactions :", &TransactionErrors{
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/transaction.go]: Clean up logs before merging

File: code/go/0chain.net/smartcontract/storagesc/handler.go [hunk 4]
```diff
@@ -94,6 +94,7 @@ func GetEndpoints(rh rest.RestHandlerI) []rest.Endpoint {
 		rest.MakeEndpoint(storage+"/replicate-authorizer-aggregates", srh.replicateAuthorizerAggregates),
 		rest.MakeEndpoint(storage+"/replicate-validator-aggregates", srh.replicateValidatorAggregates),
 		rest.MakeEndpoint(storage+"/replicate-user-aggregates", srh.replicateUserAggregates),
+		rest.MakeEndpoint(storage+"/transaction-errors", srh.getTransactionErrors),
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/handler.go]: Let's enable this in development mode only.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/sharder/handler.go [hunk 5]
```diff
@@ -261,3 +265,37 @@ func SharderStatsHandler(ctx context.Context, r *http.Request) (interface{}, err
...
+
+func TransactionErrorWriter(w http.ResponseWriter, r *http.Request) {
+
+	logging.Logger.Debug("Starting")
+	sc := GetSharderChain()
+	c := sc.Chain
+
+	edb := c.GetEventDb()
+
+	logging.Logger.Debug("jayash edb", zap.Any("edb", edb))
```
Comment (dabasov) [inline code/go/0chain.net/sharder/handler.go]: remove log

File: code/go/0chain.net/sharder/handler.go [hunk 6]
```diff
@@ -261,3 +265,37 @@ func SharderStatsHandler(ctx context.Context, r *http.Request) (interface{}, err
 		MeanScanBlockStatsTime: cc.BlockSyncTimer.Mean() / 1000000.0,
 	}, nil
 }
+
+func TransactionErrorWriter(w http.ResponseWriter, r *http.Request) {
+
+	logging.Logger.Debug("Starting")
+	sc := GetSharderChain()
```
Comment (dabasov) [inline code/go/0chain.net/sharder/handler.go]: can be done in one line `GetSharderChain().Chain.GetEventDb()`

File: code/go/0chain.net/smartcontract/dbs/event/transaction.go [hunk 7]
```diff
@@ -134,3 +144,48 @@ func (edb *EventDb) GetTransactionsForBlocks(blockStart, blockEnd int64) ([]Tran
...
+
+func (edb *EventDb) UpdateTransactionErrors() error {
+
+	// created_at for last day from now
+	lastDay := time.Now().AddDate(0, 0, -1)
+	// convert to string
+	lastDayString := lastDay.Format("2006-01-02 15:04:05")
+
+	// read all the transactions from the transaction table where status is 2 till last day
+	rows, err := edb.Get().Model(&Transaction{}).Select("output_hash, transaction_output, transaction_type").Where("status = ? and created_at > ?", 2, lastDayString).Group("output_hash, transaction_output, transaction_type").Rows()
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/transaction.go]: you can insert results of the query like
```sql
INSERT INTO transaction_errors (a, b, c)
SELECT (a, b, c) FROM transactions WHERE status =2 ...
```

**Review by dabasov — CHANGES_REQUESTED** [review]

**Review by peterlimg — APPROVED** [review]
lgtm

**Review by dabasov — APPROVED** [review]
LGTM

**Pull request merged**